### ISSUE TITLE
train: a Lightning-style trainer, define-trainer with hooks, log!, loggers and callbacks

### ISSUE THREAD

**Issue by bkc39** [bkc39/rktorch#142]
Every example hand-rolls the same loop: seed, build the model and optimizer, iterate the loader, `zero-grads!`, loss, `backward!`, `step!`, print a loss, sometimes save a checkpoint or a sample grid (05-mnist, 06-gpt, 07-asr, 08-diffusion, and the ad-hoc training script for the diffusion run). PyTorch users reach for Lightning for exactly this: the model declares what a step is, a trainer owns the loop, and logging, checkpointing and device placement are shared machinery. Do the same, in the house style of `define-layer` and `define-dataset`: a form whose clauses are the hooks, over a small protocol.

## Proposed surface

```racket
(define-trainer diffusion-trainer (net sched)
  #:init (#:base [base 32])
  (set! net (UNet #:base base))
  (set! sched (linear-schedule))
  #:optimizers
  (adam (parameters net) #:lr 2e-4)
  #:training-step (batch step)
  (define-values (xb _yb) batch)
  (define t ...) (define noise ...)
  (define loss (mse-loss (net (q-sample sched xb t noise) t) noise))
  (log! 'loss loss)
  loss
  #:validation-step (batch step)
  ...
  #:on-epoch-end (epoch)
  (log! 'samples (sample net sched 64)))

(fit (diffusion-trainer #:base 128) train-loader
     #:val val-loader
     #:epochs 500
     #:device (accelerator-if-available)
     #:logger (csv-logger "train.log")
     #:callbacks (list (checkpoint-every 25 "unet.safetensors")
                       (early-stopping 'val-loss #:patience 10)))
```

- Hooks, named as Lightning names them, all optional except `#:training-step`: `#:training-step`, `#:validation-step`, `#:test-step`, `#:optimizers` (one or several, with optional schedulers), `#:on-train-start`, `#:on-epoch-start`, `#:on-epoch-end`, `#:on-batch-end`. A hook body sees the fields in scope like a `#:forward` body. `#:training-step` returns the loss; the trainer does the zero, backward and step.
- `log!` inside a hook records a scalar (or a tensor, image grid, text) under a key; the trainer reduces scalars per epoch (mean by default) and hands everything to the logger. Loggers are a small generic: a stdout logger printing one line per epoch with the running loss and time, a CSV/JSONL file logger, and room for a TensorBoard writer later. Messages, not only numbers: `(log-message! "wrote ~a" path)` reaches the same sinks.
- `fit` owns the loop: seeding, moving the model to the device, `train!`/`eval!` around validation, `with-no-grad` for validation, gradient clipping and accumulation as options, the callbacks (checkpoint, early stopping, learning-rate schedule), and a resumable state (epoch, optimizer state, RNG) in the checkpoint.
- Datasets and loaders come from #87's `dataloader`; the model is any `layer?`, so nothing in `torch/nn` changes.

## Where it lives

`torch/train/` (`trainer.rkt` for the form and `fit`, `logging.rkt` for `log!` and the loggers, `callbacks.rkt`), with the clause grammar shared through `torch/private/definer.rkt` as `define-dataset` does. The examples then shrink to model plus hooks, and the diffusion training script becomes a `define-trainer` with an `#:on-epoch-end` that writes the grid.

Order: after the diffusion arc (#84) and alongside or after the layers home (#141), since the trainer is the third consumer of the shared definer grammar and a `#lang torch` (#136) would put `torch/train` in scope by default.

Refs: #84, #87, #136, #141.
